### ISSUE TITLE
Traitlets AttributeError: 'super' object has no attribute '_static_custom_path_default'

### ISSUE THREAD

**Issue by aihardin** [jupyter/notebook#2410]
Hi! I just updated to Jupyter 5, jupyterhub 0.6.1 and it broke my setup with a two errors when logging in. 
Spawning jupyterhub-singleuser --user=aaron --port=60827 --cookie-name=jupyter-hub-token-aaron --base-url=/user/aaron --hub-host= --hub-prefix=/hub/ --hub-api-url=http://127.0.0.1:8081/hub/api --ip=127.0.0.1
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/traitlets/traitlets.py", line 528, in get
    value = obj._trait_values[self.name]
KeyError: 'static_custom_path'

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/bin/jupyterhub-singleuser", line 297, in <module>
    main()
  File "/usr/local/bin/jupyterhub-singleuser", line 293, in main
    return SingleUserNotebookApp.launch_instance()
  File "/usr/local/lib/python3.5/dist-packages/jupyter_core/application.py", line 267, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/traitlets/config/application.py", line 657, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-7>", line 2, in initialize
  File "/usr/local/lib/python3.5/dist-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/notebook/notebookapp.py", line 1296, in initialize
    self.init_webapp()
  File "/usr/local/bin/jupyterhub-singleuser", line 268, in init_webapp
    super(SingleUserNotebookApp, self).init_webapp()
  File "/usr/local/lib/python3.5/dist-packages/notebook/notebookapp.py", line 1093, in init_webapp
    self.jinja_environment_options
  File "/usr/local/lib/python3.5/dist-packages/notebook/notebookapp.py", line 167, in __init__
    default_url, settings_overrides, jinja_env_options)
  File "/usr/local/lib/python3.5/dist-packages/notebook/notebookapp.py", line 220, in init_settings
    static_custom_path=jupyter_app.static_custom_path,
  File "/usr/local/lib/python3.5/dist-packages/traitlets/traitlets.py", line 556, in __get__
    return self.get(obj, cls)
  File "/usr/local/lib/python3.5/dist-packages/traitlets/traitlets.py", line 535, in get
    value = self._validate(obj, dynamic_default())
  File "/usr/local/bin/jupyterhub-singleuser", line 239, in _static_custom_path_default
    path = super(SingleUserNotebookApp, self)._static_custom_path_default()
AttributeError: 'super' object has no attribute '_static_custom_path_default'
```

**Comment by minrk** [comment]
To work with notebook 5.0, jupyterhub-singleuser must be upgraded to at least 0.7:

    pip install --upgrade jupyterhub

**Comment by aihardin** [comment]
That is how I upgraded jupyterhub initially. To fix my issue I had to run it again and then jupyterhub upgrade-db.

**Comment by ZelphirKaltstahl** [comment]
@minrk _"at least"_ as in _"`0.7.x` is guaranteed to work."_ or as in _"It might work with `0.7.0`, but if not, try an even newer version."_ ? I am working with a slightly modified version forked from `0.7.0`, and I get that same error, when having notebook servers `5.0.0`. Afaik I never modified anything with regard to notebook format or communication protocol in my modified JH.

**Comment by minrk** [comment]
@ZelphirKaltstahl The `_static_custom_path_default` code does not occur in [jupyterhub-singleuser in 0.7.0](https://github.com/jupyterhub/jupyterhub/blob/0.7.0/jupyterhub/singleuser.py). If you have the exact same traceback, then the fact that the error is in `/usr/local/bin/jupyterhub-singleuser` itself (as opposed to `site-packages/jupyterhub/singleuser.py`) suggests that you still have the `jupyterhub-singleuser` entrypoint from an earlier installation. You might try removing and reinstalling JupyterHub (and take care to note where it installs jupyterhub-singleuser).

**Comment by ZelphirKaltstahl** [comment]
@minrk Great help, thanks, will check that out.

**Comment by ZelphirKaltstahl** [comment]
@minrk I checked:

    diff <0.7.0 file> <my current file>

shows 0 differences. Seems I already have the `0.7.0` version of the file in my JupyterHub. When I run my JupyterHub, I actually create a Docker container and run that, so all the files in the running JupyterHub come from when I built the Docker image. I built the image today and did not touch that `singleuser.py` file in months. It shows no differences to the file you linked with the GNU `diff` tool.

Is there something else, that could be the cause of this?

**Comment by minrk** [comment]
> Is there something else, that could be the cause of this?

I don't believe so. If you have jupyterhub 0.7.0 or greater, the `jupyterhub-singleuser` script is only a few lines:

```python
#!/prefix/bin/python

from jupyterhub.singleuser import main

if __name__ == '__main__':
    main()
```

So a traceback in that file means you have `jupyterhub-singleuser` from pre-0.7. It's possible that you have installed the current jupyterhub, but the old `jupyterhub-singleuser` script is still present and still being launched. You might try searching for it in your container:

```bash
find / -name jupyterhub-singleuser
# or
which -a jupyterhub-singleuser
```

**Comment by ZelphirKaltstahl** [comment]
@minrk I misread your previous comment, I'll have to check again, thanks.

**Comment by ZelphirKaltstahl** [comment]
My  `jupyterhub-singleuser` file is under `./jupyterhub/scripts/jupyterhub-singleuser` in the jupyterhub folder.
it has the following content:

    #!/usr/bin/env python3

    from jupyterhub.singleuser import main

    if __name__ == '__main__':
        main()

My error is precisely the following:

    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 528, in get
        value = obj._trait_values[self.name]
    KeyError: 'static_custom_path'

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/usr/local/bin/jupyterhub-singleuser", line 297, in <module>
        main()
      File "/usr/local/bin/jupyterhub-singleuser", line 293, in main
        return SingleUserNotebookApp.launch_instance()
      File "/opt/conda/lib/python3.6/site-packages/jupyter_core/application.py", line 267, in launch_instance
        return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
      File "/opt/conda/lib/python3.6/site-packages/traitlets/config/application.py", line 657, in launch_instance
        app.initialize(argv)
      File "<decorator-gen-7>", line 2, in initialize
      File "/opt/conda/lib/python3.6/site-packages/traitlets/config/application.py", line 87, in catch_config_error
        return method(app, *args, **kwargs)
      File "/opt/conda/lib/python3.6/site-packages/notebook/notebookapp.py", line 1296, in initialize
        self.init_webapp()
      File "/usr/local/bin/jupyterhub-singleuser", line 268, in init_webapp
        super(SingleUserNotebookApp, self).init_webapp()
      File "/opt/conda/lib/python3.6/site-packages/notebook/notebookapp.py", line 1093, in init_webapp
        self.jinja_environment_options
      File "/opt/conda/lib/python3.6/site-packages/notebook/notebookapp.py", line 167, in __init__
        default_url, settings_overrides, jinja_env_options)
      File "/opt/conda/lib/python3.6/site-packages/notebook/notebookapp.py", line 220, in init_settings
        static_custom_path=jupyter_app.static_custom_path,
      File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 556, in __get__
        return self.get(obj, cls)
      File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 535, in get
        value = self._validate(obj, dynamic_default())
      File "/usr/local/bin/jupyterhub-singleuser", line 239, in _static_custom_path_default
        path = super(SingleUserNotebookApp, self)._static_custom_path_default()
    AttributeError: 'super' object has no attribute '_static_custom_path_default'

So somehow another `jupyterhub-singleuser` script is created inside the docker container at `/usr/local/bin/jupyterhub-singleuser`. Is this part of the installation with `pip install -e .` inside the JupyterHub directory?

In my `Dockerfile` I am doing:

    WORKDIR /srv/
    COPY jupyterhub /srv/jupyterhub
    WORKDIR /srv/jupyterhub/
    RUN python setup.py js \
     && pip install -e .

Which copies the JupyterHub from the Docker context into the Docker image and then installs JupyterHub in it. Inside that copied JupyterHub is where I checked for the content of `jupyterhub-singleuser`, which I described above.

However, I am clueless at this moment, where the `/usr/local/bin/jupyterhub-singleuser` comes from.

**Comment by minrk** [comment]
Can you share the rest of your Dockerfile? What's your base image (`FROM ...`)? If you are installing jupyterhub from 0.7, then the file with these contents won't be created, so it is presumably leftover from a previous installation. Since I see from your traceback that you are using conda in `/opt/conda` and single-user is installed at `/usr/local/bin/`, installing jupyterhub won't go in `/usr/local`, it will go somewhere like `/opt/conda/bin/jupyterhub-singleuser`. The commands I pasted above:

```bash
find / -name jupyterhub-singleuser
# or
which -a jupyterhub-singleuser
```

should point to the possibility of more than one install. You can also try removing `/usr/local/bin/jupyterhub-singleuser` prior to installing JupyterHub, to make sure this file isn't present.

**Comment by ZelphirKaltstahl** [comment]
@minrk I'll see what I can do. I might also try to get the newest stable running in my setup. It is difficult to know if the changes I have in the 0.7.0 affect this or not. In theory they should not, since I never touched anything related to notebook format though. When I have more information, I'll post.

**Comment by ZelphirKaltstahl** [comment]
@minrk I figured out where the old file came from. I was downloading the script of `0.6.1` in my Dockerfile with `wget`. That download was copied from the JupyterHub Docker image Dockerfile from DockerHub months ago and since I did not have such file in my Docker context (what I mean by that is "files next to the Dockerfile, which are copied into the Docker image") when building the Notebook Docker images, I never saw that file … Now updated that download and will see how it goes.
